python/ray/_raylet.pyx

```diff
@@ -2093,38 +2159,83 @@ cdef class CoreWorker:
                 with nogil:
                     check_status(
                         CCoreWorkerProcess.GetCoreWorker().SealReturnObject(
-                            return_id, return_ptr[0]))
+                            return_id, return_ptr[0], generator_id))
             return True
         else:
             with nogil:
                 success = (CCoreWorkerProcess.GetCoreWorker()
-                           .PinExistingReturnObject(return_id, return_ptr))
+                           .PinExistingReturnObject(
+                                   return_id, return_ptr, generator_id))
             return success
 
-    cdef store_task_outputs(
-            self, worker, outputs, const c_vector[CObjectID] return_ids,
-            c_vector[shared_ptr[CRayObject]] *returns):
+    cdef store_task_outputs(self,
+                            const CAddress &caller_address,
+                            worker, outputs,
+                            is_dynamic,
+                            const CObjectID &outer_id,
+                            c_vector[c_pair[CObjectID, shared_ptr[CRayObject]]]
+                            *returns):
         cdef:
             CObjectID return_id
             size_t data_size
             shared_ptr[CBuffer] metadata
             c_vector[CObjectID] contained_id
             int64_t task_output_inlined_bytes
+            int64_t num_returns = -1
+            shared_ptr[CRayObject] *return_ptr
+        if is_dynamic:
+            # The task specified a dynamic number of return values. Determine
+            # the expected number of return values.
+            if returns[0].size() > 0:
+                # We are re-executing the task. We should return the same
+                # number of objects as before.
+                num_returns = returns[0].size()
+            else:
+                # This is the first execution of the task, so we don't know how
+                # many return objects it should have yet.
+                num_returns = -1
+        else:
+            # The task specified how many return values it should have.
+            num_returns = returns[0].size()
 
-        if return_ids.size() == 0:
+        if num_returns == 0:
             return
 
-        n_returns = return_ids.size()
-        returns.resize(n_returns)
         task_output_inlined_bytes = 0
+        i = -1
         for i, output in enumerate(outputs):
-            if i >= n_returns:
+            if num_returns >= 0 and i >= num_returns:
                 raise ValueError(
                     "Task returned more than num_returns={} objects.".format(
-                        n_returns))
+                        num_returns))
+            while i >= returns[0].size():
```

Comment:
Shouldn't it be `if` instead of `while`?

Reply:
What about this? Is it possible that this loop will be executed more than once?

Reply:
I think it's okay either way; this way is a bit more defensive to me.